### ISSUE TITLE
Exception when exportin CSV

### ISSUE THREAD

**Issue by mgwerder** [sdaps/sdaps#175]
I am using sdaps Version 1.9.6-1 on Ubuntu. The system has no missing patches.
sdaps is installed from http://ppa.launchpad.net/benjamin-sipsolutions/sdaps-stable/ubuntu.

When doing a CSV export a python stack is dumped:
`# sdaps csv export .`

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/sdaps/__init__.py", line 65, in main
    return cmdline['_func'](cmdline)
  File "/usr/lib/python3/dist-packages/sdaps/script.py", line 118, in decorated_function
    result = function(cmdline)
  File "/usr/lib/python3/dist-packages/sdaps/cmdline/csvdata.py", line 119, in csvdata
    csvoptions=csvoptions)
  File "/usr/lib/python3/dist-packages/sdaps/csvdata/__init__.py", line 32, in csvdata_export
    survey.questionnaire.csvdata.open_csv(csvfile, image_writer, export_images, export_question_images, export_quality, csvoptions)
  File "/usr/lib/python3/dist-packages/sdaps/csvdata/buddies.py", line 50, in open_csv
    header.extend(qobject.csvdata.export_header())
  File "/usr/lib/python3/dist-packages/sdaps/csvdata/buddies.py", line 101, in export_header
    header = [ self.obj.id_csv() + '_review' ]
  File "/usr/lib/python3/dist-packages/sdaps/model/questionnaire.py", line 162, in id_csv
    if self.var:
AttributeError: 'Head' object has no attribute 'var'

Any hint how to cleverly proceed?

**Comment by benzea** [comment]
Are you able to provide the corresponding LaTeX file?

**Comment by benzea** [comment]
Hrm, right, it probably also is a project migration issue :-/

**Comment by benzea** [comment]
```
diff --git a/sdaps/model/questionnaire.py b/sdaps/model/questionnaire.py
index 167ef6a..3a83b91 100644
--- a/sdaps/model/questionnaire.py
+++ b/sdaps/model/questionnaire.py
@@ -187,6 +187,11 @@ class QObject(buddy.Object, DataObject):
         return None
 
     def __setstate__(self, data):
+        # The variable property was renamed to var at some point
+        if 'variable' in data:
+            data['var'] = data['variable']
+            del data['variable']
+
         self.__dict__ = data
         self.id = tuple(self.id)
 
```

Could you try the above patch? If not, I might actually want to see the project in question or at least some parts of the serialised json.

**Comment by mgwerder** [comment]
Patch results in same stacktrace.

# sdaps csv export /opt/examinator/infsec/Lektion10/sdaps.work
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/sdaps/__init__.py", line 65, in main
    return cmdline['_func'](cmdline)
  File "/usr/lib/python3/dist-packages/sdaps/script.py", line 118, in decorated_function
    result = function(cmdline)
  File "/usr/lib/python3/dist-packages/sdaps/cmdline/csvdata.py", line 119, in csvdata
    csvoptions=csvoptions)
  File "/usr/lib/python3/dist-packages/sdaps/csvdata/__init__.py", line 32, in csvdata_export
    survey.questionnaire.csvdata.open_csv(csvfile, image_writer, export_images, export_question_images, export_quality, csvoptions)
  File "/usr/lib/python3/dist-packages/sdaps/csvdata/buddies.py", line 50, in open_csv
    header.extend(qobject.csvdata.export_header())
  File "/usr/lib/python3/dist-packages/sdaps/csvdata/buddies.py", line 101, in export_header
    header = [ self.obj.id_csv() + '_review' ]
  File "/usr/lib/python3/dist-packages/sdaps/model/questionnaire.py", line 162, in id_csv
    if self.var:
AttributeError: 'Head' object has no attribute 'var'

Regarding JSON/project it contains a lot of data I would prefer not pasting here. Is it OK if I send it by mail?

**Comment by benzea** [comment]
Hrm. Right, the problem is related to the review comments that I added. That, in combination with a lack of an auto-assigned `var` for the header, though I really don't understand why that would happen.

```
diff --git a/sdaps/model/questionnaire.py b/sdaps/model/questionnaire.py
index 167ef6a..9f7fb37 100644
--- a/sdaps/model/questionnaire.py
+++ b/sdaps/model/questionnaire.py
@@ -187,6 +187,9 @@ class QObject(buddy.Object, DataObject):
         return None
 
     def __setstate__(self, data):
+        if 'var' not in data:
+            data['var'] = None
+
         self.__dict__ = data
         self.id = tuple(self.id)
 
```

Should work around it, I think. Though I am really not sure why this is happening in the first place.

**Comment by mgwerder** [comment]
Aaaaand horraayyy... thanks heaps. This is working.

Why this is happening? Maybe I got an explanation. I am using SDAPS for frequent examination in lessons. Between stamping of the exam and the review may be up to 4 weeks (and thus one or more version; I usually prepare exams for 2-3 weeks in one batch). Maybe that explains it. If it comes to the worst I am using SDAPS on a project directory created half a year ago.

I am using a very handy process. Last semester I corrected more than 1000 frequent tests using SDAPS. I have a well automated processing queue taking CSV exports from SDAPS, do the grading, create corrected PDF sheets, send out test results, and collate the results of all tests of a student in a summarizing report.

One last question: When will this fix (and #173) be in the stable repo?

**Comment by benzea** [comment]
This fix will not go into the repository ;-)

I think I need to re-consider the logic and load defaults before restoring the object state. That will fix both this and #173 properly without having to special case it. As to when I might do the next release, not idea right now.
